homeassistant/components/modem_callerid/sensor.py

```diff
@@ -28,33 +25,48 @@
     }
 )
 
+SERVICE_SCHEMA = vol.Schema({vol.Optional(ATTR_ENTITY_ID): cv.comp_entity_ids})
+
+_LOGGER = logging.getLogger(__name__)
 
-def setup_platform(hass, config, add_entities, discovery_info=None):
-    """Set up modem caller ID sensor platform."""
 
-    name = config.get(CONF_NAME)
-    port = config.get(CONF_DEVICE)
+async def async_setup_entry(hass, entry, async_add_entities):
+    """Set up the Modem Caller ID sensor."""
+    name = entry.data[CONF_NAME]
+    device = entry.data[CONF_DEVICE]
+    api = hass.data[DOMAIN][entry.entry_id][DATA_KEY_API]
+    async_add_entities(
+        [
+            ModemCalleridSensor(
+                hass,
+                api,
+                name,
+                device,
+                entry.entry_id,
+            )
+        ]
+    )
 
-    modem = bm(port)
-    if modem.state == modem.STATE_FAILED:
-        _LOGGER.error("Unable to initialize modem")
-        return
+    platform = entity_platform.current_platform.get()
 
-    add_entities([ModemCalleridSensor(hass, name, port, modem)])
+    platform.async_register_entity_service(
+        SERVICE_REJECT_CALL, SERVICE_SCHEMA, "reject_call"
+    )
 
 
 class ModemCalleridSensor(SensorEntity):
     """Implementation of USB modem caller ID sensor."""
 
-    def __init__(self, hass, name, port, modem):
+    def __init__(self, hass, api, name, device, server_unique_id):
         """Initialize the sensor."""
         self._attributes = {"cid_time": 0, "cid_number": "", "cid_name": ""}
-        self._name = name
-        self.port = port
-        self.modem = modem
+        self._device_class = None
+        self.device = device
+        self.api = api
         self._state = STATE_IDLE
-        modem.registercallback(self._incomingcallcallback)
-        hass.bus.listen_once(EVENT_HOMEASSISTANT_STOP, self._stop_modem)
+        self._name = name
+        self._server_unique_id = server_unique_id
+        api.registercallback(self._incomingcallcallback)
```

Comment:
```suggestion


    async def async_added_to_hass(self):
        """Called when the entity is added to Home Assistant."""
        self.api.registercallback(self._incomingcallcallback)
```

Reply:
Make sure this is ok to run in the event loop (does no I/O)

Reply:
```Logger: aiohttp.server
Source: /usr/local/lib/python3.8/site-packages/aiohttp/web_protocol.py:393
First occurred: 6:03:45 PM (1 occurrences)
Last logged: 6:03:45 PM

Unhandled exception
Traceback (most recent call last):
  File "/usr/local/lib/python3.8/asyncio/base_events.py", line 1155, in _sendfile_fallback
    await proto.drain()
  File "/usr/local/lib/python3.8/asyncio/base_events.py", line 226, in drain
    await fut
asyncio.exceptions.CancelledError

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/local/lib/python3.8/site-packages/aiohttp/web_protocol.py", line 485, in start
    resp, reset = await task
  File "/usr/local/lib/python3.8/site-packages/aiohttp/web_protocol.py", line 440, in _handle_request
    reset = await self.finish_response(request, resp, start_time)
  File "/usr/local/lib/python3.8/site-packages/aiohttp/web_protocol.py", line 591, in finish_response
    await prepare_meth(request)
  File "/usr/local/lib/python3.8/site-packages/aiohttp/web_fileresponse.py", line 241, in prepare
    return await self._sendfile(request, fobj, offset, count)
  File "/usr/local/lib/python3.8/site-packages/aiohttp/web_fileresponse.py", line 96, in _sendfile
    await loop.sendfile(transport, fobj, offset, count)
  File "/usr/local/lib/python3.8/asyncio/base_events.py", line 1131, in sendfile
    return await self._sendfile_fallback(transport, file,
  File "/usr/local/lib/python3.8/asyncio/base_events.py", line 1161, in _sendfile_fallback
    await proto.restore()
  File "/usr/local/lib/python3.8/asyncio/base_events.py", line 263, in restore
    self._transport.resume_reading()
  File "/usr/local/lib/python3.8/asyncio/sslproto.py", line 344, in resume_reading
    self._ssl_protocol._transport.resume_reading()
AttributeError: 'NoneType' object has no attribute 'resume_reading'
```
I get this every time a call starts and ends.